### PR TITLE
Navigation: Fix click-button size, submenu directions, scrollbars.

### LINKED ISSUE #206: Visible horisontal scroll bar
**Describe the bug**
At some screen sizes there is a visible horisontal scroll bar.

**To Reproduce**
1. Go to the front of the website
2. Reduce the browser window width to somewhere between 800 and 900px.
3. See the scroll bar

**Expected behavior**
I expect there to not be a visible horisontal scroll bar.

**Screenshots**
visible scroll bar at the bottom of the browser window

**Device:**
 - Device: PC
 - OS: Windows10
 - Browser: Chrome

**Additional context**
I believe there was a discussion in the Gutenberg GitHub repository about a default "reset" CSS for themes, but 
I can't find it now and whatever the result was, it did not include overflow.
Considering adding `overflow-x: hidden;` to the body element for the front only, in style.css

### PR DESCRIPTION
**Pull request by jasmussen**
## Description

Fixes https://github.com/WordPress/twentytwentytwo/issues/206. As reported there, there are a few issues with dropdown submenus and their nested versions as well:

<strike>- Even when the submenu is closed, its contents take up space and can in some circumstances create a horizontal scrollbar.</strike>
- Submenus should open leftwards when right aligned. This worked, but not at small breakpoints.
<strike>- Not reported in that ticket, but surfaced by it — menus set to open on click had smaller buttons for the submenus than for hover.</strike>
- The contents of burger menus should always be top aligned.

Some of the bugs originally fixed by this PR have been fixed separately. But the two remaining issues are still good to land.

## How has this been tested?

<strike>
- Create navigation menus that features submenus, and sub sub menus (nested menus).
- Test with and without click enabled.
- Test a justified-right submenu on small (<782) and bigger breakpoints. They should both open leftwards, but <782 menus should open leftwards and _downwards_ (as shown in the GIF above).
- Bonus: use the web inspector to verify that submenus are zero width even when not shown.

Note that we use `visibility: hidden:` to hide submenus so we can nicely fade them in and out. `display: none;` is an alternative that won't let us animate it as nicely.</strike>

Test instructions and GIFs are shown in https://github.com/WordPress/gutenberg/pull/36215#issuecomment-966174501.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: packages/block-library/src/navigation/style.scss
Comment (jasmussen) [inline packages/block-library/src/navigation/style.scss]: This overflow rule I believe I had in the past, and then removed it. In this PR it appears to work as intended, with no ill side effects. But I would appreciate thorough testing of submenus, with and without click, and with and without keyboard usage. Just to make sure I'm not missing anything.

**Review by vcanales — COMMENTED** [review]
The right alignment fix works but only when explicitly aligned right; it doesn't seem to fix the problem when the navigation menu is used in a template, where it happens to sit at the right of the screen.

Kapture 2021-11-04 at 13 45 13

It's also noticeable that long submenu items might overflow outside of the viewport; is it possible to wrap them conditionally? ie. when they exceed the viewport.

image

Other than that, I didn't find any other details! 👍

**Comment by jasmussen** [comment]
Thanks so much for the thorough review! I'll look at addressing this.

**Comment by jasmussen** [comment]
> The right alignment fix works but only when explicitly aligned right; it doesn't seem to fix the problem when the navigation menu is used in a template, where it happens to sit at the right of the screen.

Yes, that's a bit of a limitation of how things work at the moment, but beyond the scope of this PR to fix. I created #36241 to track this further.

**Comment by jasmussen** [comment]
> It's also noticeable that long submenu items might overflow outside of the viewport; is it possible to wrap them conditionally? ie. when they exceed the viewport.

It appears I intentionally added nowrap to menu items in https://github.com/WordPress/gutenberg/pull/35077. I'll see if I can remove it again, but I think it was necessary due to some scaling issues.

It also surfaced some other issues, which I'll look into.

**Comment by jasmussen** [comment]
Fixed two issues. First off, the explicit application of a z-index when the mobile responsive menu is enabled caused this:
bug

Setting it to "auto" fixed it:
bugfix

I also removed the `nowrap`, because even without it, I can no longer reproduce the issue I saw in https://github.com/WordPress/gutenberg/pull/35077

**Comment by jasmussen** [comment]
Alright, fixed a last issue with justificaitons broken when mobile responsiveness was enabled.

Screenshot 2021-11-05 at 09 27 29
 
This one should be good to go now.

**Comment by jasmussen** [comment]
This one needs a good rebase after the merging #36169.

**Comment by vcanales** [comment]
> This one needs a good rebase after the merging #36169.

By itself, it's looking good to me. Rebase-wise, there was only one conflict to fix:

```css
.wp-block-navigation__responsive-container-content {
<<<<<<< HEAD
  display: contents;
=======
  display: flex;
>>>>>>> 329693fea1 (Fix issue with justification and responsive container.)
}
```

`display: flex` was replaced by `display: contents` in #36169. I'll push the rebased branch in order to move forward :)

**Comment by jasmussen** [comment]
Thank you for the rebase help. It might need more changes, though, as 36169 removed the justification classes, so menus no longer open in the correct direction at all. I think @tellthemachines had some ideas to explore here as well.

**Comment by jasmussen** [comment]
I've extracted some of the small fixes (line height, overflow etc) into #36298, which can likely land easily as it's small. This PR should rebase right over it, but is still blocked on either restoring the justification classes or finding another way forward. But by extracting the other things, this one can focus up a bit.

**Comment by jasmussen** [comment]
Alright, I think this one is ready for review again. A lot of what this PR originally fixed has now been made moot as it was extracted and fixed separately, so there are primarily two things left to test, and both in my own testing appear fixed by this PR.

1. With a justified-right navigation menu that has multiple levels of nesting, verify that menus on desktop breakpoints open down and leftwards. On mobile breakpoints, they open directly downwards:
resizing

2. With a justified-right (or centered) navigation menu set to always be burgered, verify that menus are always aligned to the top, even if they are left/centered/right justified:
Screenshot 2021-11-11 at 11 07 17

**Comment by tellthemachines** [comment]
@jasmussen I can no longer reproduce either of those issues on trunk. To clarify, were the submenus on small breakpoints not opening to the bottom? I've been comparing trunk to this branch and not quite sure what has changed 😅

**Comment by jasmussen** [comment]
Sorry the editing of the original issue actually removed helpful instructions. I still see this happening on trunk:



That's becasue the rules for justifying the menus are wrapped in a media query. So the steps to test:

- Create navigation menus that features submenus, and sub sub menus (nested menus).
- Set them to open on click, set them justified right.
- Open the submenu 2 levels deep
- Resize the window smaller than 782px. 

At around that breakpoint, the menus should in this PR start to open _downwards_, but still be right aligned to their parent item. In trunk, they just open rightwards at that point.

**Comment by tellthemachines** [comment]
Oooh got it, thanks. I was looking at the nested submenus and ignoring where they were sitting at the top level 🤦 

I think this is good to go!

**Review by tellthemachines — APPROVED** [review]

**Comment by jasmussen** [comment]
Thank you!

**Pull request merged**
